### DISCUSSION TITLE
Raycast against terrain3D at -x yields no/error results?

### DISCUSSION THREAD

**Discussion by andreasng** [TokisanGames/Terrain3D#548] [Questions] [answered]
I admit i'm not a coder, in any shape or form, but i have this raycast here(which "works", but not...):

```
func raycast_to_world(start_position: Vector3) -> Vector3:
	var ray_params = PhysicsRayQueryParameters3D.new()
	ray_params.from = start_position + Vector3(.0, rayHeight, .0)
	ray_params.to = start_position + Vector3(.0, -rayHeight, .0)
	ray_params.hit_back_faces = true #-----------tried this doesnt make a difference
	ray_params.hit_from_inside = true #----------tried this doesnt make a difference
	ray_params.collide_with_areas = false
	ray_params.collide_with_bodies = true
	ray_params.collision_mask = 0b00000000_00000000_00001000_00000000 #12 - my terrain mask id

	var space_state = get_world_3d().direct_space_state
	var result = space_state.intersect_ray(ray_params)
	
	#print(result) #<--------------------this will crash the game if raycasting against negative x

	if result:
		return result.position
	else:
		print("raycast failed!")
```

And if I raycast at ie. position 4,-2,-8 its fine. also at .2,17,500....
But if i get into the negative x positions it yields nothing. 

Does this make sense because the code is rubbish, or is it a temporary bug in the system?

**Comment by TokisanGames** [comment]
> print(result) 

This shouldn't crash ever, no matter the value. If so, that's a godot bug and nothing to do with Terrain3D.

Is collision enabled? 

Do you have regions defined where you are raycasting?

Did you read this? 
https://terrain3d.readthedocs.io/en/stable/docs/collision.html

Vertical raycasting can be entirely replaced with `get_height()`. Non-vertical can be replaced with `get_intersection()` though see #511 and  https://github.com/TokisanGames/Terrain3D/issues/511#issuecomment-2424140081

**Reply by andreasng** [reply]
yes i am as confused as ever. I run around on the entire terrain just fine. but this is just puzzling. 
get height is for terrain? I'll try that (and probably want it instead!)

**Reply by andreasng** [reply]
oh....
"This function works fine if called only once per frame, such as for a mouse pointer detecting terrain position. More than once per frame will produce conflicts."
i planned on doing this ... alot...

**Reply by andreasng** [reply]
Query Many Heights is for me then

**Reply by TokisanGames** [reply]
> I run around on the entire terrain just fine. 

If you're using this for a player you should use physics move_and_slide, just like in our demo.

If you want to do a vertical raycast, you should use get_height instead.

Raycasts do work fine when properly setup in code, within regions, with collision.

**Comment by andreasng** [comment]
ALRIGHT!!!!1

after extensive testing i have now found a very very idiotic solution to this. And i will retain that either godot or terrain3d has a flaw in how raycasting collides with it. Let me explain:

This will fail. and crash if you query result.collider.name forexample

```
func raycast_to_world(start_position: Vector3) -> Vector3:
	print("raycast origin: ", start_position)
	var ray_params = PhysicsRayQueryParameters3D.new()
	ray_params.from = start_position + Vector3(0.0, rayHeight, 0.0)
	ray_params.to = start_position - Vector3(0.0, rayHeight, 0.0)
	ray_params.collide_with_bodies = true
	ray_params.collision_mask = 0b00000000_00000000_00001000_00000000 #12 - terrain
	var result = get_world_3d().direct_space_state.intersect_ray(ray_params)
	print(result)
```

BUT... i work with unreal daily at work, and i came to think about a thing unreal had where rays that were clean traces (ie not a sphere trace or a box trace, but just a laser pointer trace) would .... "seep".... through ......................"cracks" in ALL collision. 

So as a hail mary i modified this little tiny tiny thing:

```
func raycast_to_world(start_position: Vector3) -> Vector3:
	print("raycast origin: ", start_position)
	var ray_params = PhysicsRayQueryParameters3D.new()
	ray_params.from = start_position + Vector3(0.1, rayHeight, 0.1)
	ray_params.to = start_position - Vector3(0.1, rayHeight, 0.1)
	ray_params.collide_with_bodies = true
	ray_params.collision_mask = 0b00000000_00000000_00001000_00000000 #12 - terrain
	var result = get_world_3d().direct_space_state.intersect_ray(ray_params)
	print(result)
```

Can you spot the difference? 

yes. The ray is not straight, and i might, for good measure, "flicker" it as well. 

**Toki**, i hope you will have a look at this and see if it is a major flaw in godot, or a little easy fix in t3d.

peace! 😆 🚀

**Comment by TokisanGames** [comment, marked as answer]
I have no problem running this in 4.3 using the demo, player.gd:
```
func _physics_process(p_delta) -> void:
	...
	print(raycast_to_world(global_position))

func raycast_to_world(start_position: Vector3) -> Vector3:
	var ray_params = PhysicsRayQueryParameters3D.new()
	ray_params.from = start_position + Vector3(.0, 1., .0)
	ray_params.to = start_position + Vector3(.0, -1000., .0)
	var space_state = get_world_3d().direct_space_state
	var result = space_state.intersect_ray(ray_params)
	print(result)
	if result:
		return result.position
	else:
		print("raycast failed!")
		return Vector3.INF
```

There's no crash if the raycast doesn't hit or in -X. If there's a region it prints the hit without issue:
```
{ "position": (-321.822, 0, 0.229156), "normal": (0, 1, 0), "face_index": -1, "collider_id": 35920020962, "collider": [Wrapped:0], "shape": 0, "rid": RID(6768868458527) }
(-321.822, 0, 0.229156)
```

If there's no region, it prints what is expected:
```
{  }
raycast failed!
(inf, inf, inf)
```
